SI-8787 Regextraction is null-proof

som-snytt opened the pull request:
A null never matches.

That's what they say at eHarmony, anyway.

som-snytt committed:
SI-8787 If you love nulls, so does Regex

Regex is robust when unapplying null.  A null never matches.

som-snytt committed:
SI-8787 Update doc for Regex

Simplify the exposition and update the examples.

Emphasize pattern matching, especially using `r.unanchored`
instead of `for (r(x,y,z) <- r findFirstIn text)`.

Certain details are moved to method docs.

It would be nice to fix matching package doc,
but the doc must attach to the package object, it seems.
Introducing a package object is not binary-compatible.

Includes a doc line edit on 2.12, anticipating the merge.

som-snytt commented:
Heh, heh. Introducing an empty package obj to carry scaladoc is not binary compat.

Ichoran commented on src/library/scala/util/matching/Regex.scala:
```diff
@@ -28,117 +27,109 @@
+ *  }}}
  *
- *  // val start = " ".r.findAllIn(msg).start // throws an IllegalStateException
+ *  To check only whether the `Regex` matches, ignoring any groups:
```
Maybe add "use a wildcard extractor" (if that is the canonical term for it), instead of just letting the code complete the sentence?

som-snytt replied:
Yup. "Sequence wildcard."

Ichoran commented on src/library/scala/util/matching/Regex.scala:
```diff
@@ -28,117 +27,109 @@
- *  Regex does not provide a method that returns a [[scala.Boolean]]. One can
- *  use [[java.lang.String]] `matches` method, or, if `Regex` is preferred,
- *  either ignore the return value or test the `Option` for emptyness. For example:
+ *  To find all matches:
```
You might want to mention here that it finds all non-overlapping matches.

som-snytt replied:
OK.  Tricky to KISS.  "See the method doc about finding matches that are overlapping." Or not exactly that.

Ichoran commented on src/library/scala/util/matching/Regex.scala:
```diff
@@ -188,6 +191,10 @@ class Regex private[matching](val pattern: Pattern, groupNames: String*) extends
    *  }
    *  val p2 = "a(b*)c".r
+   *  val p2Matches = "abbbc" match {
+   *    case p2(_*) => true
```
This works, but a single _ is enough since there is only a single capturing group.

som-snytt replied:
Yes, I'll show `(_), (_*) and (foo @ _*)` styles.

Ichoran commented on src/library/scala/util/matching/Regex.scala:
```diff
@@ -274,30 +287,47 @@ class Regex private[matching](val pattern: Pattern, groupNames: String*) extends
+   *  }}}
+   *
+   *  To return overlapping matches, it is possible to formulate a regular expression
+   *  that does not consume the overlapping region.
```
Dropping the phrase "lookahead" is probably a good idea given how hard it is to search for symbols like ?=

som-snytt replied:
Dropping as in name dropping, right.

Ichoran commented on src/library/scala/util/matching/Regex.scala:
```diff
@@ -588,7 +630,7 @@ object Regex {
 
-  /** Provides information about a succesful match.
+  /** Provides information about a successful match.
    */
```
Since you're revising this so extensively, keeping the close on an extra line doesn't help anything.

som-snytt replied:
Sincere thx for the close reading.

som-snytt committed:
SI-8787 Addressing feedback, additional periods.

gkossakowski commented:
@som-snytt, @lchoran, what's the status of this PR?

som-snytt commented:
I addressed the comments.  They don't call him "Regex" Kerr for nothing.  I ran the doc and looked it over for obvious formatting glitches.  A comb would no doubt turn up improvements, which could be incremental or this PR could block on further eyeballs.

Ichoran commented:
LGTM

gkossakowski commented:
Thanks!
